core/tcp_callback_export.go

```diff
@@ -109,6 +99,26 @@ func tcpRecvFn(arg unsafe.Pointer, tpcb *C.struct_tcp_pcb, p *C.struct_pbuf, err
 		}
 	}
 
+	select {
+	case buf, ok := <-readCh:
+		if !ok {
+			C.tcp_recved(tpcb, p.tot_len)
+			C.tcp_shutdown(tpcb, 1, 0)
+			return C.ERR_OK
+		}
+		if len(buf) < int(p.tot_len) {
```

Comment:
This will break if the app passes a small buffer.

Instead of failing, why don't you copy as much as the buffer can hold, calling `C.tcp_recved(tpcb, len(buf))`?

Reply:
Calling `tcp_recved` only cause lwip to advertise a larger TCP window size, there is no way to partly receive a pbuf passed to `tcpRecvFn`, you either accept a whole pbuf and return ERR_OK, or reject it and return some other errors such as ERR_CONN, lwip would hold the pbuf for the next `tcpRecvFn` call in the latter case.

Reply:
I worry this new behavior would break any client that passes a buffer smaller than the TCP MSS, which seems to be the size of the pbufs. Actually it's worse: it will break if the buffer is shorter than the pbuf chain, since you're using the `p.tot_len`, not `p.len`.

I see you have to [consume a full pbuf in order to return OK](https://www.nongnu.org/lwip/2_0_x/raw_api.html). In that case you could iterate until a full pbuf is read, then return OK.

Reply:
Then the blocking issue is re-introduced.